flamingpy/noise/cv.py

```diff
@@ -0,0 +1,623 @@
+        if isinstance(code, EGraph):
+            self.egraph = code
+        else:
+            self.code = code
+            self.egraph = code.graph
+
+        # Set some egraph properties to attributes of self.
+        # (Note: self._adj also generates indices in the EGraph)
+        self._adj = self.egraph.adj_generator(sparse=True)
+        self._N = len(self.egraph)
+        self._to_points = self.egraph.to_points
+
+        # Get indices of nodes we wish to designate as ideal.
+        perfect_points = self.egraph.graph.get("perfect_points")
+        if perfect_points is None:
+            self._perfect_inds = None
+        else:
+            self._perfect_inds = [self.egraph.to_indices[point] for point in perfect_points]
+
+        # Set some noise model parameters.
+        self.delta = delta
+        self.p_swap = kwargs.get("p_swap")
+        self.states = kwargs.get("states") or {"p": np.empty(0, dtype=int)}
+        self.sampling_order = kwargs.get("sampling_order")
+
+        # Assign state labels.
+        self.populate_states()
+
+    # Error correction methods
+    def apply_noise(self, sampling_order=None, rng=default_rng()):
+        """Apply cv-level noise to the graph state.
+
+        Identify the nodes of the EGraph with CV states, measure the syndrome
+        assuming the specified noise model, and convert the simulated homodyne
+        outcomes to bit values.
+
+        This method modifies the attributes of self.egraph.
+        """
+        if sampling_order is None:
+            sampling_order = self.sampling_order
+        self.measure_syndrome(sampling_order=sampling_order, rng=rng)
+        self.inner_decoder()
+
+    def populate_states(self, rng=default_rng()):
+        """Populate the graph state with state labels.
+
+        Assume the graph states consists of a combination of squeezed states
+        and GKP+ states. A non-zero self.p_swap overrides indices specified in
+        self.states and uses a binomial distribution to identify some indices
+        as p-squeezed states.
+
+        Raise an exception if both p_swap and p indices are supplied.
+
+        This method modifies self.egraph.
+        """
+        # Generate indices of squeezed states based on swap-out
+        # probability self.p_swap, if supplied.
+        if self.p_swap:
+            self._generate_squeezed_indices(rng)
+
+        # Associate remaining indices with GKP states.
+        self._generate_gkp_indices()
+
+        # Give the EGraph nodes state attributes.
+        self._apply_state_labels()
+
+    def measure_syndrome(self, sampling_order="initial", rng=default_rng()):
+        """Measure the syndrome for memory-mode error correction of the grpah
+        states."""
+        return self.measure_hom(
+            quad="p", inds=self.code.all_syndrome_inds, sampling_order=sampling_order, rng=rng
+        )
+
+    def inner_decoder(self, translator=GKP_binner):
```

Comment:
Again, translator should be part of the `__init__`. I know, this will make it such that it has many arguments. This is a hint that it should be split in smaller objects, as discussed before.

Reply:
Good point -- as with comment above, I agree.